components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt

```diff
@@ -16,41 +17,22 @@ class ToolbarPresenter(
     private val toolbar: Toolbar,
     private val sessionManager: SessionManager,
     private val sessionId: String? = null
-) : SessionManager.Observer, Session.Observer {
-    private var activeSession: Session? = null
+) : SessionObserver(sessionManager) {
 
     /**
      * Start presenter: Display data in toolbar.
      */
     fun start() {
-        activeSession = sessionId?.let {
-            sessionManager.findSessionById(sessionId)
-        } ?: run {
-            sessionManager.register(this)
-            sessionManager.selectedSession
-        }
-
-        activeSession?.register(this)
+        val session = sessionId?.let { sessionManager.findSessionById(sessionId) }
+        session?.let { observe(it) } ?: run { observeSelected() }
         initializeView()
     }
 
-    /**
-     * Stop presenter from updating the view.
-     */
-    fun stop() {
-        sessionManager.unregister(this)
-        activeSession?.unregister(this)
-    }
-
     /**
      * A new session has been selected: Update toolbar to display data of new session.
      */
     override fun onSessionSelected(session: Session) {
-        activeSession?.unregister(this)
-
-        activeSession = session
-        session.register(this)
-
+        super.onSessionSelected(session)
         initializeView()
```

Comment:
This `super` call is something that is easy to forget and feels like it is not the best solution yet. We could try to at least annotate `SessionObserver.onSessionSelected()` with [@CallSuper](https://developer.android.com/reference/android/support/annotation/CallSuper). This will show a warning in the IDE if you forget that. Haven't tried whether that works in Kotlin though.

Reply:
Yes, good idea, `CallSuper` works fine. I've added it. This override isn't actually needed for any of our samples, as we're always starting/stopping the fragment and never actually have to observe `onSessionSelected`. I opted for keeping it in `ToolbarPresenter` anyway as it should work regardless. Wdyt?